### DISCUSSION TITLE
vips_tonelut/vips_maplut to tone map an image

### DISCUSSION THREAD

**Discussion by kipcole9** [libvips/libvips#3549] [Help wanted]
I probably have another conceptual disconnect here but I'm trying to tone map an image and I've been assuming that the behaviour would be similar to applying a curve in Lightroom.

Using the example below I was expected that the tonal range would be severely compressed but the result I'm seeing is basically no different to the source image.

```bash
% wget https://upload.wikimedia.org/wikipedia/commons/c/cb/Broadway_tower_edit.jpg
# Tone curve that compresses the tonal range and reduces the highlights by the max -30
% vips tonelut lut.vips --Lb 40 --Lw 50 -H -30
# The resulting image is visually no different to the source image
% vips maplut Broadway_tower_edit.jpg mapped.jpg lut.vips
```

What is the appropriate way to use `vips_tonemap`/`vips_maplut` ?

**Comment by kipcole9** [comment]
Seems I need to set `in-max` and `out-max` to the maximum value of the band format of the image (so 255 for a UCHAR image) and I start to see adjustments. I'm also `cast`ing the LUT to the same band format as the source image - is that typically the right thing to do?

One more supplemental question:  If I set `Lb` and `Lw` but do not set `H`/`M`/`S` I don't see the range compression I was expecting. I only see the LUT being applied if some combination of `H`/`M`/`S` are set.  Is that to be expected?  If there another way to compress the tone range (ie applying a new black point and white point)?

**Comment by kipcole9** [comment]
Is it correct to say that `Ps`/`Pm`/`Ph` are expressed as fractions in the range `Lb`..`Lw`?  That is, do they refer to the range between the black point and the white point. Or do they refer to the range of the original image?

**Comment by jcupitt** [comment]
I'd try in nip2, you can drag the sliders and see what all the settings do:

![image](https://github.com/libvips/libvips/assets/580843/d40c4a4b-c449-4942-ae71-70616c802218)

**Comment by kipcole9** [comment]
Thanks John. Sadly nip2 still doesn't play well with MacOS so it's a non-starter for me.  

If you set the sliders for black point 40, white point 60 (values don't matter really other than to compress the range) and leave all the adjustments at 0 then I expected the image to have its range compressed. My findings are that there are no adjustments made.

**Reply by jcupitt** [reply]
It works OK for me:

![image](https://github.com/libvips/libvips/assets/580843/4dda9053-d3a0-4026-aff0-9a64ff9f5ced)

That's this one:

https://github.com/libvips/nip2/releases/download/v8.6.0-beta2/nip2-8.6.0-beta2.app.dmg

On an M2 mac mini.

**Reply by jcupitt** [reply]
> My findings are that there are no adjustments made.

That's right, you use Ps, Pm, Ph to set the positions of the shadow, midpoint and highlight, then S, M and H to push them up or down.

**Reply by kipcole9** [reply]
Oh, exciting. My home brew version doesn't - will definitely try the release version from GitHub.

**Reply by kipcole9** [reply]
> > My findings are that there are no adjustments made.
> 
> 
> 
> That's right, you use Ps, Pm, Ph to set the positions of the shadow, midpoint and highlight, then S, M and H to push them up or down.

Hmm, then I'm not sure what the black point and white point do - perhaps they just set the baseline from which the shadow, mid and nightlight points are calculated?

**Reply by jcupitt** [reply]
That's right. Try in nip2, it's super obvious how it works.

**Comment by ewelot** [comment]
I like the visualization but I am unable to create it by myself. I cannot figure out how to make the graph of mapping values ... :(
Could you share the nip2 workspace, please?

Thomas

**Reply by jcupitt** [reply]
Hi @ewelot, it's all built in, just load an image and click Toolkits / Image / Level / Tone curve.

The nip2 menus are hard to find things in -- if you click View / Toolkit browser there's another interface to the same thing with keyword search. Try entering eg. "tone".

**Reply by ewelot** [reply]
That is what I have used so far, but I don't find a way to show the graph of mapping values ...
I am using nip2 version 8.7.1.

**Reply by jcupitt** [reply]
You should see a thumbnail of the graph in the workspace (like in the screenshot), then doubleclick that to get a pop up graph.

If you don't see a thumbnail, your nip2 has been built without graphing support. You'll need to rebuild it against libgoffice-0.8.

**Reply by ewelot** [reply]
I don't see a thumbnail. Thanks for the hint regarding the build of the program. I'll habe a look into this ...
